### PR TITLE
SchedulerPredicates [Serial] validates resource limits of pods that are allowed to run - set resource requests as well

### LINKED ISSUE #87191: SchedulerPredicates [Serial] validates resource limits of pods that are allowed to run - possibly invalid test
**What happened**:

I am questioning validity of the test. Comments under `test/e2e/scheduling/predicates.go` of `validates resource limits of pods that are allowed to run` say:
```
// This test verifies we don't allow scheduling of pods in a way that sum of
// limits of pods is greater than machines capacity.
// It assumes that cluster add-on pods stay stable and cannot be run in parallel
// with any other test that touches Nodes or Pods.
// It is so because we need to have precise control on what's running in the cluster.
// Test scenario:
// 1. Find the amount CPU resources on each node.
// 2. Create one pod with affinity to each node that uses 70% of the node CPU.
// 3. Wait for the pods to be scheduled.
// 4. Create another pod with no affinity to any node that need 50% of the largest node CPU.
// 5. Make sure this additional pod is not scheduled.
/*
	Release : v1.9
	Testname: Scheduler, resource limits
	Description: Scheduling Pods MUST fail if the resource limits exceed Machine capacity.
*/
```

The test scenario described as it is sounds reasonable. The only nitpick I have is the very first sentence saying "`This test verifies we don't allow scheduling of pods in a way that sum of limits of pods is greater than machines capacity`". Pod's resource limits are ignored in predicates. They take effect in the scoring phase when priorities are computed. Also, priorities are computed only when there are at least two candidate nodes. So after revisiting the step number 5 it's impossible to make sure the additional pod does or does not get scheduled when setting the resource limits. There are either no candidate nodes or one. In which case the limits are ignored and the test passes in the first case and fails in the second. In case there are at least two candidate nodes, one of them will always get picked. In which case the additional pod gets scheduled (no matter what the resource limits field says).

In the framework:
- resource limits are accessed through `ResourceLimits.PostFilter` method [1]
- `ResourceLimits.PostFilter` is used to perform pre-scoring [2]
- in the scoring phase score is computed through `ResourceLimits.Score` method [3]

The test is alternating between passing and failing state. It passes when predicates give empty list of candidate nodes and failing when the list of candidate nodes is non-empty. Thus, making the pod resource limits ineffective.

[1] https://github.com/kubernetes/kubernetes/blob/be26fbc638ff5ed952ccae70e68baa712b3c60c8/pkg/scheduler/framework/plugins/noderesources/resource_limits.go#L141-L153
[2] https://kubernetes.io/docs/concepts/configuration/scheduling-framework/#post-filter
[3] https://github.com/kubernetes/kubernetes/blob/be26fbc638ff5ed952ccae70e68baa712b3c60c8/pkg/scheduler/framework/plugins/noderesources/resource_limits.go#L106-L125

**What you expected to happen**:

Have another node schedule the additional pod validating that all saturated nodes (with `filler-pods-*` scheduled) have lower score (0 in case of `ResourceLimits` plugin). Or, invalidate the test by removing it.

Can anyone check the test validity and confirm/disprove my statement about ineffectiveness of the resource limits?

**How to reproduce it (as minimally and precisely as possible)**:
Make sure there is at least one node with non-zero free cpu resource before running the step 5.

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`): master HEAD
- Cloud provider or hardware configuration: N/A
- OS (e.g: `cat /etc/os-release`): N/A
- Kernel (e.g. `uname -a`): N/A
- Install tools: N/A
- Network plugin and version (if this is a network-related bug): N/A
- Others: N/A

### PR DESCRIPTION
**Pull request by ingvagabund**
**What type of PR is this?**
/kind flake

**What this PR does / why we need it**:
NodeResourceFit plugin's Filter method responsible for checking if a pod fits a given node ignores resource limits and acknowledge resource requests only. Given both tests validating resource limits of pods were setting only pod resource limits, ability of NodeResourceFit plugin to properly filter nodes was not tested at all.

**Which issue(s) this PR fixes**:

Fixes #87191

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

cc @kubernetes/sig-scheduling-test-failures

### CODE REVIEW

**Comment by ahg-g** [comment]
can you please include a link to a failed run?

File: test/e2e/scheduling/predicates.go [hunk 1]
```diff
@@ -18,9 +18,10 @@ package scheduling
 
 import (
 	"fmt"
-	podutil "k8s.io/kubernetes/pkg/api/v1/pod"
 	"time"
 
+	podutil "k8s.io/kubernetes/pkg/api/v1/pod"
+
```
Comment (alculquicondor) [inline test/e2e/scheduling/predicates.go]: put it in the same block as the next imports

**Comment by ingvagabund** [comment]
> can you please include a link to a failed run?

I can't find any upstream run where the test failed. Though, we experience the failure in OpenShift from time to time: https://bugzilla.redhat.com/show_bug.cgi?id=1779303#c2. It either passes since there are no candidate nodes or fails when there's one. Is there any upstream e2e run with serial tests where I can take a look and validate?

**Comment by ahg-g** [comment]
> > can you please include a link to a failed run?
> 
> I can't find any upstream run where the test failed. Though, we experience the failure in OpenShift from time to time: https://bugzilla.redhat.com/show_bug.cgi?id=1779303#c2. It either passes since there are no candidate nodes or fails when there's one. Is there any upstream e2e run with serial tests where I can take a look and validate?

Here: https://testgrid.k8s.io/sig-release-master-blocking#Conformance%20-%20GCE%20-%20master&include-filter-by-regex=resource&include-filter-by-regex=SchedulerPredicates

I am looking at previous successful https://storage.googleapis.com/kubernetes-jenkins/logs/ci-kubernetes-gce-conformance-latest/1217212937886765059/build-log.txt

If you look for "additional-pod", you will find the test in the log. The pod is actually getting "FailedScheduling" event due to "Insufficient cpu", which I still don't understand why.

**Comment by ingvagabund** [comment]
@ahg-g thanks for the links.

> If you look for "additional-pod", you will find the test in the log. The pod is actually getting "FailedScheduling" event due to "Insufficient cpu", which I still don't understand why.

Even after summing up all the cpu requests and going through the math done in the test, all three nodes should still have available cpu to run the additional-pod (even after adding +100m for pods with undefined cpu resource request). Would be helpful to see what's actually inside scheduler's node cache. To see if there are any pods left that are not supposed to be there.

**Comment by ahg-g** [comment]
> Even after summing up all the cpu requests and going through the math done in the test, all three nodes should still have available cpu to run the additional-pod. Would be helpful to see what's actually inside scheduler's node cache. To see if there are any pods left that are not supposed to be there.

but it doesn't matter what's in the scheduler cache, the pod is supposedly requesting 0 CPU, but is getting rejected for "Insufficient CPU" on all three worker nodes, which shouldn't happen.

**Comment by ingvagabund** [comment]
> If you look for "additional-pod", you will find the test in the log. The pod is actually getting "FailedScheduling" event due to "Insufficient cpu", which I still don't understand why.

@ahg-g can we merge this PR and investigate the case separately?

**Comment by alculquicondor** [comment]
If not set, I believe kube-apiserver adds a request of 100 mCPU per container.

**Comment by ahg-g** [comment]
> FWIIW, If not set, I believe kube-apiserver adds a request of 100 mCPU per container.

can you link to the code?

**Comment by ingvagabund** [comment]
> FWIIW, If not set, I believe kube-apiserver adds a request of 100 mCPU per container.

I got confused by this as well. Kube-scheduler only adds +100m when priorities are computed. It does not effect predicates.

**Pull request merged**